src/matching/quoteMatcher.ts

```diff
@@ -0,0 +1,44 @@
+/**
+ * QuoteMatcher matches quoted strings, respecting escaped quotes (\") and friends
+ */
+export class QuoteMatcher {
+  static escapeChar = "\\";
+
+  private char: string;
+  private start: number;
+  private corpus: string;
+  private quoteMap: boolean[] = [];
+
+  constructor(char: string, corpus: string, start: number) {
+    this.char = char;
+    this.start = start;
+    this.corpus = corpus;
+
+    // Loop over corpus, marking quotes and respecting escape characters.
+    for (let i = 0; i < corpus.length; i++) {
+      if (corpus[i] === QuoteMatcher.escapeChar) {
+        i += 1;
+        continue;
+      }
+      this.quoteMap[i] = corpus[i] === this.char;
+    }
+  }
+
+  findOpening(): number {
+    // First, search backwards to see if we could be inside a quote
```

Comment:
If we pass `start` to this method like `findClosing`, we don't need to store `start` as well, make sense?